src/control/common/pci_utils.go

```diff
@@ -37,3 +42,289 @@ func ParsePCIAddress(addr string) (dom, bus, dev, fun uint64, err error) {
 
 	return
 }
+
+// PCIAddress represents a standard PCI address with domain and BDF.
+type PCIAddress struct {
+	Domain   string
+	Bus      string
+	Device   string
+	Function string
+}
+
+func (pa *PCIAddress) String() string {
+	if pa == nil {
+		return ""
+	}
+
+	return fmt.Sprintf("%s:%s:%s.%s", pa.Domain, pa.Bus, pa.Device, pa.Function)
+}
+
+// Equals compares string representation of address.
+func (pa *PCIAddress) Equals(o *PCIAddress) bool {
+	if pa == nil || o == nil {
+		return false
+	}
+
+	return o.String() == pa.String()
+}
+
+// IsVMDBackingAddress indicates whether PCI address is a VMD backing device.
+func (pa *PCIAddress) IsVMDBackingAddress() bool {
+	if pa == nil {
+		return false
+	}
+
+	return pa.Domain != "0000"
+}
+
+// NewPCIAddress creates a PCIAddress stuct from input string.
+func NewPCIAddress(addr string) (*PCIAddress, error) {
+	dom, bus, dev, fun, err := parsePCIAddress(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	pa := &PCIAddress{
+		Bus:      fmt.Sprintf("%02x", bus),
+		Device:   fmt.Sprintf("%02x", dev),
+		Function: fmt.Sprintf("%01x", fun),
+	}
+
+	if dom == 0 {
+		pa.Domain = "0000"
+	} else {
+		pa.Domain = fmt.Sprintf("%x", dom)
+	}
+
+	return pa, nil
+}
+
+// PCIAddressList represents a list of PCI addresses.
+type PCIAddressList []*PCIAddress
+
+// Contains returns true if provided address is already in list.
+func (pal *PCIAddressList) Contains(a *PCIAddress) bool {
```

Comment:
If you decide to keep this as a slice type alias, most of these `PCIAddressList` methods should not have a pointer receiver. Only the methods which modify the length/capacity of the slice (`Add()`, `AddStrings()`) need a pointer receiver in order to allow `append()` to change the underlying slice data structure. The rest of the methods should be non-pointer receivers, and this will allow you to simplify the code so that you don't need to check for nil pointers or return pointers to slices.